src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs

```diff
@@ -13,50 +14,97 @@ namespace Microsoft.Extensions.Caching.Memory
     internal sealed partial class CacheEntry : ICacheEntry
     {
         private static readonly Action<object> ExpirationCallback = ExpirationTokensExpired;
+        private static readonly AsyncLocal<CacheEntry> _current = new AsyncLocal<CacheEntry>();
 
         private readonly MemoryCache _cache;
 
         private CacheEntryTokens _tokens; // might be null if user is not using the tokens or callbacks
-        private TimeSpan? _absoluteExpirationRelativeToNow;
-        private TimeSpan? _slidingExpiration;
-        private long? _size;
+        private TimeSpan _absoluteExpirationRelativeToNow;
+        private TimeSpan _slidingExpiration;
+        private long _size = -1;
         private CacheEntry _previous; // this field is not null only before the entry is added to the cache and tracking is enabled
         private object _value;
-        private CacheEntryState _state;
+        private DateTime _absoluteExpiration;
+        private short _absoluteExpirationOffsetMinutes;
+        private bool _isDisposed;
+        private bool _isExpired;
+        private bool _isValueSet;
+        private byte _evictionReason;
+        private byte _priority = (byte)CacheItemPriority.Normal;
 
         internal CacheEntry(object key, MemoryCache memoryCache)
         {
             Key = key ?? throw new ArgumentNullException(nameof(key));
             _cache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
-            _previous = memoryCache.TrackLinkedCacheEntries ? CacheEntryHelper.EnterScope(this) : null;
-            _state = new CacheEntryState(CacheItemPriority.Normal);
+            if (memoryCache.TrackLinkedCacheEntries)
+            {
+                var holder = _current;
+                _previous = holder.Value;
+                holder.Value = this;
+            }
         }
 
-        /// <summary>
-        /// Gets or sets an absolute expiration date for the cache entry.
-        /// </summary>
-        public DateTimeOffset? AbsoluteExpiration { get; set; }
+        // internal for testing
+        internal static CacheEntry Current => _current.Value;
 
-        /// <summary>
-        /// Gets or sets an absolute expiration time, relative to now.
-        /// </summary>
-        public TimeSpan? AbsoluteExpirationRelativeToNow
+        private ref ulong RawAbsoluteExpiration => ref Unsafe.As<DateTime, ulong>(ref _absoluteExpiration);
```

Comment:
Is this safe / ok?

Reply:
Looks like @stephentoub had a similar comment:

> General feedback is I'd like to avoid using Unsafe.As between DateTime and long in code outside of corelib.

Reply:
It's safe - I don't see how DateTime layout could significantly change.

Reply:
Can it just use `.Ticks`? How much of a perf win does this have over using Ticks?